### PR TITLE
[MYEN-373] add login in to cache server

### PR DESCRIPTION
**Pull request by dwojno**
- [x] iam-client-lib logges in to iam-cache-server
- [x] iam-client-lib request data using authentication with JWT token
- [x] iam-client-lib asks for only one signing message to get public key and generate identity proof
- [x] refactor auth module
- [x] iam-client-lib handles non authorized errors

### CODE REVIEW

File: src/iam/iam-base.ts [hunk 1]
```diff
@@ -199,98 +195,127 @@ export class IAMBase {
     this.setupENS();
   }
 
-  private async setupProvider({
+  private async initSigner({
```
Comment (jrhender) [inline src/iam/iam-base.ts:198]: Nice rename 👍

File: src/iam/iam-base.ts [hunk 2]
```diff
@@ -199,98 +195,127 @@ export class IAMBase {
...
-        },
-        infuraId: this._connectionOptions.infuraId,
-        bridge: this._connectionOptions.bridgeUrl
-      });
-      await this._walletConnectProvider.enable();
-      this._signer = new providers.Web3Provider(this._walletConnectProvider).getSigner();
-      const publicKey = await this.getPublicKey();
-      this._didSigner = new IdentityOwner(this._signer, this._provider, publicKey);
+      const provider = new providers.Web3Provider(metamaskProvider);
+      this._signer = provider.getSigner();
```
Comment (jrhender) [inline src/iam/iam-base.ts:250]: I find this part a bit confusing as I think the outcome is that `this._signer` is associated with the Web3Provider but then we pass a different (JsonRpc) provider into the `IdentityOwner` constructor. So `IdentityOwner` is then kind of associated with two different providers 🤔 ? I think that could be a source of confusion as to which network/chain "Switchboard" is actually connected to. Not super related to this PR and I don't have a great solution, just bringing it up for thought 😅

File: src/iam/iam-base.ts [hunk 3]
```diff
@@ -199,98 +195,127 @@ export class IAMBase {
...
+        iss: `did:ethr:${address}`,
+        claimData: {
+          blockNumber: await this._provider?.getBlockNumber()
+        }
+      };
+
+      const encodedPayload = base64url(JSON.stringify(payload));
+      return { encodedHeader, encodedPayload };
+    }
+    throw Error("Signer not initialized");
```
Comment (jrhender) [inline src/iam/iam-base.ts:286]: I would use the same `ERROR_MESSAGES.SIGNER_NOT_INITIALIZED` here

File: src/iam/iam-base.ts [hunk 4]
```diff
@@ -199,98 +195,127 @@ export class IAMBase {
...
     }
-    if (this._runningInBrowser) {
-      const metamaskProvider = (await detectMetamask({
-        mustBeMetaMask: true
-      })) as any;
-      if (metamaskProvider && useMetamask) {
-        if (initializeMetamask) {
-          await metamaskProvider.request({
+    if (!this._runningInBrowser) {
+      throw new Error("Initialization of signer not possible in nodejs env without private key ");
```
Comment (jrhender) [inline src/iam/iam-base.ts:213]: I would add a new entry to `ERROR_MESSAGES` to use here.

**Review by jrhender — APPROVED** [review]
A couple comments to add to use `ERROR_MESSAGES` but otherwise looks good.

File: src/cacheServerClient/cacheServerClient.ts [hunk 5]
```diff
@@ -76,6 +77,11 @@ export class CacheServerClient implements ICacheServerClient {
     });
   }
 
+  async login(claim: string) {
+    const { data } = await this.httpClient.post<{ token: string }>("/login", { claim });
+    this.httpClient.defaults.headers.common["Authorization"] = `Bearer ${data.token}`;
```
Comment (jrhender) [inline src/cacheServerClient/cacheServerClient.ts:82]: Another aspect I thought of is whether or not this is the way that the jwt should be stored. I think this is safest spot as it in browser memory (safer than localStorage or cookie, but has downside that don't persist on new tab or refresh). Is this your understanding as well @dwojno @manihagh ?

Comment (dwojno) [reply src/cacheServerClient/cacheServerClient.ts:82]: @jrhender yes, I agree we should add option to store jwt in the cookie as configured option, but still we need to have an option to return token for nodejs apps to be able to login

File: src/iam/iam-base.ts (hunk 4 above)
Comment (dwojno) [reply src/iam/iam-base.ts:213]: 👍

File: src/iam/iam-base.ts (hunk 3 above)
Comment (dwojno) [reply src/iam/iam-base.ts:286]: 👍

File: src/cacheServerClient/cacheServerClient.ts (hunk 5 above)
Comment (jrhender) [reply src/cacheServerClient/cacheServerClient.ts:82]: @dwojno Ah, sorry, I wasn't trying to say that we necessarily need an option to store jwt as a cookie. I think the way that it is currently being done in this PR is better than cookie or localStorage for our requirements. I just wanted to confirm that this was your and Mani's view as well.

**Pull request merged**